src/integration/integration.service.ts

```diff
@@ -0,0 +1,407 @@
+    private readonly configService: ConfigService,
+    private readonly storeService: CommercetoolsService,
+    private readonly voucherifyService: VoucherifyService,
+  ) {
+    this.storeService.setCartUpdateListener(
+      (cart, storeActions, helperToGetProductsFromStore) =>
+        this.validateCouponsAndGetAvailablePromotions(
+          cart,
+          storeActions,
+          helperToGetProductsFromStore,
+        ),
+    );
+    this.storeService.setOrderRedeemListener((order) =>
+      this.redeemVoucherifyCoupons(order),
+    );
+  }
+
+  public async validateCouponsAndGetAvailablePromotions(
+    cart: Cart,
+    storeData?: StoreData,
+    helperToGetProductsFromStore?: any,
+  ): Promise<undefined> {
+    const { id, customerId, anonymousId, sessionKey, coupons, items } = cart;
+    let uniqCoupons: Coupon[] = uniqBy(coupons, 'code');
+    if (coupons.length !== uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'COUPONS: Duplicates found and deleted',
+      });
+    }
+
+    const { promotions, availablePromotions } =
+      await this.voucherifyService.getPromotions(cart, uniqCoupons);
+
+    if (!uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'No coupons applied, skipping voucherify call',
+      });
+
+      if (isClass(storeData)) {
+        storeData.setAvailablePromotions(availablePromotions);
+      }
+      return;
+    }
+
+    const deletedCoupons = uniqCoupons.filter(
+      (coupon) => coupon.status === 'DELETED',
+    );
+
+    deletedCoupons
+      .filter((coupon) => coupon.type !== 'promotion_tier')
+      .map((coupon) =>
+        this.voucherifyConnectorService.releaseValidationSession(
+          coupon.code,
+          sessionKey,
+        ),
+      );
+
+    if (deletedCoupons.length === uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'Deleting coupons only, skipping voucherify call',
+      });
+
+      if (typeof storeData?.setAvailablePromotions === 'function') {
+        storeData.setAvailablePromotions(availablePromotions);
+      }
+      return;
+    }
+
+    this.logger.debug({
+      msg: 'Attempt to apply coupons',
+      coupons: uniqCoupons,
+      id,
+      customerId,
+      anonymousId,
+    });
+
+    uniqCoupons = filterCouponsByLimit(
+      uniqCoupons,
+      getCouponsLimit(
+        this.configService.get<number>('COMMERCE_TOOLS_COUPONS_LIMIT'),
+      ),
+    );
+
+    let validatedCoupons: ValidationValidateStackableResponse =
+      await this.voucherifyConnectorService.validateStackableVouchers(
+        buildValidationsValidateStackableParamsForVoucherify(
+          uniqCoupons.filter((coupon) => coupon.status != 'DELETED'),
+          cart,
+          mapItemsToVoucherifyOrdersItems(items),
+        ),
+      );
+
+    const inapplicableCoupons = [];
+    if (validatedCoupons.valid === false) {
+      getCouponsByStatus(validatedCoupons.redeemables, 'INAPPLICABLE').forEach(
```

Comment:
you can move it above `if` block

```ts
const inapplicableCoupons = getCouponsByStatus(validatedCoupons?.redeemables || [], 'INAPPLICABLE');
if(validatedCoupons.valid === false){ // or !validatedCoupons.valid
 // shorter if -> easier to understand
}

```

it also will help you to keep `inapplicableCoupons` type (currently it's `any`)